aesd-char-driver/aesd-circular-buffer.c

```diff
@@ -77,39 +74,45 @@ struct aesd_buffer_entry *aesd_circular_buffer_find_entry_offset_for_fpos(struct
 * Any necessary locking must be handled by the caller
 * Any memory referenced in @param add_entry must be allocated by and/or must have a lifetime managed by the caller.
 */
-void aesd_circular_buffer_add_entry(struct aesd_circular_buffer *buffer, const struct aesd_buffer_entry *add_entry)
+const char* aesd_circular_buffer_add_entry(struct aesd_circular_buffer *buffer, const struct aesd_buffer_entry *add_entry)
 {
-    /**
-    * TODO: implement per description
-    */
-       // Check if inputs are valid pointers
-       if (buffer == NULL || add_entry == NULL) {
-        return;
+    const char* ret_val = NULL;
+
+    if (buffer == NULL || add_entry == NULL)
+        return NULL;
+
+    // Save pointer to the buffer being replaced if full
+    if (!buffer->full) {
+        buffer->entry[buffer->in_offs] = *add_entry;
+        buffer->in_offs = (buffer->in_offs + 1) % AESDCHAR_MAX_WRITE_OPERATIONS_SUPPORTED;
     }
     
-    // Store current state before modification
-    size_t current_in = buffer->in_offs;
-    
-    // Calculate next input position
-    buffer->in_offs = (buffer->in_offs + 1) % AESDCHAR_MAX_WRITE_OPERATIONS_SUPPORTED;
-    
-    // Copy the entry to the buffer at the current input position
-    buffer->entry[current_in] = *add_entry;
-    
-    // Determine if we need to update out_offs based on buffer state
-    if (buffer->full) {
-        // Buffer was already full, so move the output pointer
-        buffer->out_offs = buffer->in_offs;
-    } else if (buffer->in_offs == buffer->out_offs) {
-        // We've just filled the buffer completely
+    else{
+               // If buffer is full, overwrite the oldest entry
+               buffer->entry[buffer->in_offs] = *add_entry;
```

Comment:
Before overwriting the oldest entry, you should set `ret_buffer` with the `buffer->entry[buffer->in_offs].buffptr` so the caller can free it.